5.1 - Lazy router scopes

markstory opened the pull request:
This is an experiment that I had on recent travels. I spent a small amount of time with SPX profiling a simple 'hello world' template render. I wanted to see if there were any opportunities for quick wins. I found some interesting overhead in route generation. We've long known that route parsing can be a non-trivial amount of framework overhead.

I used routes from an application I have. It includes 65 routes organized into 10 scopes. I was interested in the performance of two functions. First, h `loadRoutes` method is responsible for loading application and plugin routes. Next, the `RoutingMiddleware::process` function is a great signal of the framework overhead, as it concludes with the controller being created and invoked. Currently in `5.x` we see:

`loadRoutes`

![before` loadroutes](https://github.com/cakephp/cakephp/assets/24086/653bacc4-c920-4cef-b37b-6f1400f6e6a1)

`RoutingMiddleware::process`
![before routingmiddleware](https://github.com/cakephp/cakephp/assets/24086/fb419654-94d1-44b5-a09e-f95f8cc98069)

My idea was that for applications that use routing scopes/prefixes/plugins we have a really simple way to defer route generation. Each of these concepts introduces a static prefix. When a request is processed we only need to load the scope that matches. The other scopes don't matter, and might not ever be needed. This could allow us to shave a measurable amount of time off the 'framework time'. 

I've implemented this theory and measured the inclusive time of `loadRoutes` and `RoutingMiddleware::process` again:

`loadRoutes`
![after loadroutes](https://github.com/cakephp/cakephp/assets/24086/e624274b-e47a-4e58-999d-5e33bb872e1c)

`RoutingMiddleware::process`
![after routingmiddleware](https://github.com/cakephp/cakephp/assets/24086/fc0f5c50-b4de-4afc-a53e-5e7c14fa678e)

By deferring route construction response times fell from 4ms to 3.2ms. The time spent in `process` went from 2.12ms to 1.6ms, which I think is decent savings.

I consider this a prototype and not yet production ready, but it 'works' and is measurably faster for the 'hello world' scenario. However, it creates some interesting tradeoffs. Those tradeoffs are:

1. While there are no 'breaks' in the API, there could be semantically different behavior in applications that access and manipulate routes directly. Applications that only use `Router` to generate URLs should not be impacted.
2. While we've optimized the simple scenario where a view generates 0 URLs, most responses *do* generate URLs. This means that when we go to generate URL strings we need to build **all** routes as `RouteCollection` doesn't have a good way to skip route scopes *yet*.
3. It is an overall increase incomplexity. This is another behavior that increases complexity with questionable value for real-world applications. Sure they may 'startup' faster, but we've not eliminated any work yet.

If we can find a reasonable solution to skipping scopes during URL matching we could deliver improved performance for more applications. One potential solution could be to use scope `params` as a way to fast-forward scopes that won't create routes that match anyways. This would work well with plugin based applications.

I also think this needs to be an opt-in change that applications enable. If it gets traction and doesn't generate many issues we can consider making it the default in the future.

So does this seem like it would be worth pursuing further?

markstory committed:
First draft of lazy routing scopes.

They aren't really that lazy yet as all scopes are resolved as soon as
a match/parseRequest() call is made. By making route scopes evaluate
lazily tests needed to be adjusted so that resolution was done.

markstory committed:
Make scopes evaluate lazily based on path matching.

This should help with application startup and dispatching when an
application has many routes that are organized with scopes.

ADmad commented:
> This means that when we go to generate URL strings we need to build all routes as RouteCollection doesn't have a good way to skip route scopes yet.

So in the "real world" scenario this change won't have any impact on the total response time until we can make changes to URL generation too right?

markstory commented:
> So in the "real world" scenario this change won't have any impact on the total response time until we can make changes to URL generation too right?

Yes. These changes would allow us to amortize the cost of building routes until later, but no actual savings for real world applications yet.

I will see how hacky parameter matching would be in `match()`. If that can be made to work we'd be able to provide savings to real world applications that use scopes + parameters.

markstory committed:
First pass at partial scope resolution during match()

Instead of having to eagerly evaluate route scopes during match()
we can lazily evaluate any routing scopes that have parameters defined.

my hope is that application developers can build more 'tiered' scopes
for plugins and controllers to better optimize and organize their
routes.

markstory commented:
I was able to implement partial scope resolution for scopes that use parameters. I think this gives some carrot to the way I think routes can be defined with simpler to read routes. Patterns like this

```php
    $builder->scope('/password', ['controller' => 'Users'], function ($builder) {
        $builder->connect('/reset', ['action' => 'resetPassword'], ['_name' => 'users:passwordReset']);
        $builder->connect('/new/{token}', ['action' => 'newPassword'], ['_name' => 'users:newPassword'])
            ->setPass(['token']);
    });
```

Provide performance improvements and make routes subjectively easier to read as you have layers of indentation in each of the major sections of your application. This also has benefits for applications composed of plugins and prefixes as these constructs include routing parameters for free. This could yield benefits for applications that put no effort into changing their routes and are already using prefixes or plugins with routes.

markstory committed:
Fix prefix matching

markstory committed:
Fix phpstan errors

markstory committed:
Make using lazy scopes opt-in

This gives us an escape hatch if for some reason lazy routes are not
working with an application.

markstory committed:
Using lazy scopes should inherit

markstory commented:
I've made lazy scopes an opt-in behavior now. I updated Router tests that use scopes to have run with both eager and lazy scope builders. Having this be an opt-in change should mitigate any breaking changes.

We could enable this by default in the `cakephp/app` for new applications, and ask application developers to give feedback on any performance gains.

markstory committed:
Fix phpcs and increase coverage

othercorey commented on src/Routing/RouteBuilder.php:
```diff
@@ -941,7 +948,33 @@ public function scope(string $path, Closure|array $params, ?Closure $callback =
+     * @param bool $value The scope evaluation mode
+     * @return $this
+     */
+    public function useLazyScopes(bool $value)
```
Should we stick with the enable/disable paradigm here?

othercorey commented on src/Routing/RouteCollection.php:
```diff
@@ -126,6 +133,91 @@ public function add(Route $route, array $options = []): void
+        if (empty($this->unresolvedScopes)) {
+            return;
+        }
+        assert($path === null || $url === null, 'Must provide one of `path` or `url`');
```
The condition doesn't match the message.

markstory committed:
Incorporate feedback
